### PR TITLE
Add `--include-internal` option to CLI tool.

### PR DESCRIPTION
**Pull request by stevepm**
- Add ability to add an internal option to all properties
- When properties are marked as internal, docs will not be generated for them unless `--include-internal` is passed via the CLI.
- Add the ability to describe why an endpoint is marked as `internal` or `nodoc!`.

### CODE REVIEW

File: lib/brainstem/cli/generate_api_docs_command.rb [hunk 1]
```diff
@@ -158,6 +158,10 @@ def option_parser
             options[:sink][:options][:output_extension] = extension
           end
 
+          opts.on('--internal', 'Generate docs for all `internal!` flagged presenters/controllers') do |_|
```
Comment (shirish-pampoorickal) [inline lib/brainstem/cli/generate_api_docs_command.rb]: I suggest naming this `--include-internal` or `--allow-internal`

File: lib/brainstem/api_docs/controller.rb [hunk 2]
```diff
@@ -101,13 +103,19 @@ def tag_groups
...
       end
 
       def valid_sorted_endpoints
         endpoints.sorted_with_actions_in_controller(const)
       end
+
+      private
+
+      def nodoc_for?(config)
+        !!(config[:nodoc] || (config[:internal] && !internal))
```
Comment (shirish-pampoorickal) [inline lib/brainstem/api_docs/controller.rb]: I would rename `internal` to be `include_internal` or something better. `config[:internal]` & `internal` is confusing

File: lib/brainstem/concerns/controller_dsl.rb [hunk 3]
```diff
@@ -65,10 +65,20 @@ def documented!
         # whereas setting it within an action context will force that action to
         # be undocumented.
         #
-        def nodoc!
+        def nodoc!(_description = nil)
           configuration[brainstem_params_context][:nodoc] = true
```
Comment (shirish-pampoorickal) [inline lib/brainstem/concerns/controller_dsl.rb]: Why are we not saving the description?

File: lib/brainstem/concerns/controller_dsl.rb [hunk 4]
```diff
@@ -65,10 +65,20 @@ def documented!
...
         end
 
+        #
+        # Specifies that the scope should not be documented unless the `--internal`
+        # flag is passed in the CLI. Setting this on the default context
+        # will force the controller to be undocumented, whereas setting it
+        # within an action context will force that action to be undocumented.
+        #
+        def internal!(_description = nil)
+          configuration[brainstem_params_context][:internal] = true
```
Comment (shirish-pampoorickal) [inline lib/brainstem/concerns/controller_dsl.rb]: Why are we not saving the description?

File: lib/brainstem/concerns/presenter_dsl.rb [hunk 5]
```diff
@@ -39,18 +39,22 @@ def associations(&block)
...
           configuration[:title] = options.merge(info: str)
         end
 
-        def description(str, options = { nodoc: false })
+        def description(str, options = { nodoc: false, internal: false })
           configuration[:description] = options.merge(info: str)
         end
 
-        def nodoc!
+        def nodoc!(_description = nil)
```
Comment (shirish-pampoorickal) [inline lib/brainstem/concerns/presenter_dsl.rb]: Why are we not saving the description?

File: lib/brainstem/concerns/presenter_dsl.rb [hunk 6]
```diff
@@ -39,18 +39,22 @@ def associations(&block)
...
           configuration[:nodoc] = true
         end
 
+        def internal!(_description = nil)
```
Comment (shirish-pampoorickal) [inline lib/brainstem/concerns/presenter_dsl.rb]: Why are we not saving the description?

File: lib/brainstem/concerns/controller_dsl.rb (hunk 3 above)
Comment (stevepm) [reply lib/brainstem/concerns/controller_dsl.rb]: we don't currently have a use for it. In the future, when we actually use the description somewhere, we can save the data and use it, but right now, there isn't any reason to. It's just setting us up for a future where we'll have this data for whatever use we decide is appropriate.

File: lib/brainstem/concerns/controller_dsl.rb (hunk 4 above)
Comment (stevepm) [reply lib/brainstem/concerns/controller_dsl.rb]: see above

File: lib/brainstem/concerns/presenter_dsl.rb (hunk 5 above)
Comment (stevepm) [reply lib/brainstem/concerns/presenter_dsl.rb]: see above

File: lib/brainstem/concerns/presenter_dsl.rb (hunk 6 above)
Comment (stevepm) [reply lib/brainstem/concerns/presenter_dsl.rb]: see above

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb [hunk 7]
```diff
@@ -43,6 +44,12 @@ def format_title!
               definition.merge! title: presenter_title(presenter)
             end
 
+            def sort_properties!
+              definition[:properties] = definition[:properties].sort.each_with_object({}) do |(key, val), obj|
```
Comment (shirish-pampoorickal) [inline lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: return if definition[:properties].blank?

File: lib/brainstem/concerns/controller_dsl.rb [hunk 8]
```diff
@@ -78,11 +88,15 @@ def nodoc!
         # Setting the +:nodoc+ option marks this presenter as 'internal use only',
         # and causes formatters to display this as not indicated.
         #
+        # Setting the +:internal+ option marks this presenter as 'internal use only',
+        # and causes formatters to not display this unless documentation is generated with
+        # `--internal` flag.
```
Comment (shirish-pampoorickal) [inline lib/brainstem/concerns/controller_dsl.rb]: `--include-internal`

File: lib/brainstem/concerns/controller_dsl.rb [hunk 9]
```diff
@@ -100,11 +114,18 @@ def presents(target_class = :default, options = { nodoc: false })
         # controller or action as nondocumentable, instead, use the
         # +.nodoc!+ method in the desired context without a block.
         #
+        # Setting the +:internal+ option marks this title as 'internal use only',
+        # and causes formatters to fall back to the controller constant or to
+        # the action name as appropriate unless documentation is generated with
+        # `--internal` flag. If you are trying to set the entire
```
Comment (shirish-pampoorickal) [inline lib/brainstem/concerns/controller_dsl.rb]: `--include-internal`

File: lib/brainstem/concerns/controller_dsl.rb [hunk 10]
```diff
@@ -115,11 +136,15 @@ def title(text, options = { nodoc: false })
         # Setting the +:nodoc+ option marks this description as 'internal use
         # only', and causes formatters not to display a description.
         #
+        # Setting the +:internal+ option marks this description as 'internal use
+        # only', and causes formatters not to display a description unless documentation
+        # is generated with `--internal` flag.
```
Comment (shirish-pampoorickal) [inline lib/brainstem/concerns/controller_dsl.rb]: `--include-internal`

**Pull request merged**